app/logical/sources/strategies/hentai_foundry.rb

```diff
@@ -0,0 +1,116 @@
+# * https://www.hentai-foundry.com/pictures/user/Afrobull/795025o
+#
+# Preview URLs
+#
+# * https://thumbs.hentai-foundry.com/thumb.php?pid=795025&size=350
+#
+# Profile URLs
+#
+# * https://www.hentai-foundry.com/user/kajinman/profile
+# * https://www.hentai-foundry.com/pictures/user/kajinman
+# * https://www.hentai-foundry.com/pictures/user/kajinman/scraps
+# * https://www.hentai-foundry.com/user/J-likes-to-draw/profile
+
+module Sources
+  module Strategies
+    class HentaiFoundry < Base
+      BASE_URL =    %r!\Ahttps?://(?:www\.)?hentai-foundry\.com!i
+      PAGE_URL =    %r!#{BASE_URL}/pictures/user/(?<artist_name>[\w-]+)/(?<illust_id>\d+)(?:/[\w-]*)?\z!i
+      PROFILE_URL = %r!#{BASE_URL}/(?:pictures/)?user/(?<artist_name>[\w-]+)(?:/[a-z]*)?\z!i
+      IMAGE_URL =   %r!\Ahttps?://pictures\.hentai-foundry\.com/\w/(?<artist_name>[\w-]+)/(?<illust_id>\d+)/[\w-]+\.\w+\z!i
+
+      def domains
+        ["hentai-foundry.com"]
+      end
+
+      def site_name
+        "Hentai Foundry"
+      end
+
+      def image_urls
+        image = page&.search("#picBox img")
+
+        return [] unless image
+
+        image.to_a.map { |img| "https:#{img["src"]}" }
+      end
+
+      def preview_urls
+        image_urls.map do |img|
+          "https://thumbs.hentai-foundry.com/thumb.php?pid=#{img[IMAGE_URL, :illust_id]}&size=250"
+        end
+      end
+
+      def page_url
+        return nil if illust_id.blank? || artist_name.blank?
+        "https://www.hentai-foundry.com/pictures/user/#{artist_name}/#{illust_id}"
+      end
+
+      def page
+        return nil if page_url.blank?
+
+        agent.get(page_url)
+      rescue Mechanize::ResponseCodeError => e
+        return nil if e.response_code.to_i == 404
+        raise
+      end
+
+      def tags
+        tags = page&.search(".boxbody [rel='tag']")&.to_a
+
+        return [] unless tags
+
+        tags.map(&:text)
+      end
+
+      def artist_name
+        urls.map { |url| url[PROFILE_URL, :artist_name] || url[PAGE_URL, :artist_name] || url[IMAGE_URL, :artist_name] }.compact.first
+      end
+
+      def profile_url
+        return nil if artist_name.blank?
+        "https://www.hentai-foundry.com/user/#{artist_name}"
+      end
+
+      def artist_commentary_title
+        page&.search("#picBox .imageTitle")&.text
+      end
+
+      def artist_commentary_desc
+        page&.search("#descriptionBox .picDescript")&.to_html
+      end
+
+      def dtext_artist_commentary_desc
+        DText.from_html(artist_commentary_desc).gsub(/\A[[:space:]]+|[[:space:]]+\z/, "").gsub(/\n+/, "\n")
+      end
+
+      def normalizable_for_artist_finder?
+        artist_name.present?
+      end
+
+      def normalized_for_artist_finder?
+        url =~ PROFILE_URL
+      end
+
+      def illust_id
+        url[PAGE_URL, :illust_id] || url[IMAGE_URL, :illust_id]
+      end
+
+      def agent
+        mech = Mechanize.new
+
+        mech.get("https://www.hentai-foundry.com/?enterAgree=1")
```

Comment:
For my personal danbooru instance I also wrote a hentai-foundry strategy (never bothered upstreaming it because it's fairly low-quality)

Here's how how I do the `page` method:
```rb
def page
  return nil unless page_url.present?
  doc = Cache.get("hf:#{page_url}", 1.minute) do 
    HTTParty.get("#{page_url}?enterAgree=1").body
  end
  Nokogiri::HTML(doc)
end
```

While it's also not optimal, here's a few points:
* it uses `?enterAgree=1` directly on the page we want instead of first visiting the root page
* it doesn't use mechanize, though it does still use httparty
  * my impression is that the `Danbooru::Http` wrapper for `httprb` seems to be the prefered http client now (?)
  * `httprb` doesn't seem to like cookies - when it follows a redirect it just seems to throw away the cookies it got
  * it uses manual caching, which seems to have mostly been removed in favour of `Danbooru::Http`, it does use caching, though, unlike yours

So I don't know what the preferred way to do it would be, but even if you keep using uncached Mechanize, you should still be able to use `?enterAgree=1` directly.

One thing I noticed, though, is
* when visiting [`https://www.hentai-foundry.com/pictures/user/Afrobull/795025?enterAgree=1`](https://www.hentai-foundry.com/pictures/user/Afrobull/795025?enterAgree=1&) (NSFW) it 301 redirects
* when visiting [`http://www.hentai-foundry.com/?enterAgree=1`](http://www.hentai-foundry.com/?enterAgree=1) it immediately responds with a 200 OK

Reply:
Yeah, the reason I didn't use ?enterAgree=1 directly on the post is because of the redirect. 
Also in some instances afterwards it redirected me to the main page instead of the post, not sure why. I can't seem to replicate it right now though. I'll try doing some mass tests and see if it was just an issue with my browser.

Using Danbooru::Http has the added benefit that fetching deleted sources correctly returns 404, unlike with mechanize, so a proper 404 error can be thrown. How would I go to allow redirections? The wrapper doesn't seem to make the [HTTP.follow](https://github.com/httprb/http/wiki/Redirects) method available.

Reply:
From what I can tell, `.get(url, follow: {max_hops: 1})` seems to work - except that as I mentioned it doesn't actually work because httprb doesn't like cookies. As far as I can tell it doesn't actually keep the cookies it got from the first response (the redirect) and makes the second one without any cookies - which obviously doesn't work.

Reply:
See https://github.com/httprb/http/issues/264

Reply:
See /httprb/http / issues/264.
I deleted the first comment because they might not like a "Add hentai-foundry support" mention on their issues but it keeps them anyways... I don't want to add another one, though

Reply:
Yeah, github saves those even if they're deleted. No way to get rid of it now.  Same reason I added "fixes" in the pull request body rather than the commit.

Anyway, yeah, I'm not sure if there's a better way to do this than your method. HTTParty also returns 404 correctly on deleted urls, so it's preferable in any case to mechanize. I'll add that and add a check for deleted illustrations.

Reply:
So you'd have to do it manually, except:
```rb
[31] pry(main)> HTTP.get("https://www.hentai-foundry.com/").cookies
=> #<HTTP::CookieJar:0x0634f398
 @store=
  #<HTTP::CookieJar::HashStore:0x0634f2f0
   @gc_index=1,
   @gc_threshold=150,
   @jar=$CENSORED,
   @logger=nil,
   @mon_data=#<Monitor:0x0634f290>,
   @mon_data_owner_object_id=74448>>
[32] pry(main)> Danbooru::Http.cache(1.minute).get("https://www.hentai-foundry.com/").cookies
ArgumentError: bad argument (expected URI object or URI string)
from /usr/lib/ruby/2.7.0/uri/common.rb:739:in `URI'
```

Reply:
I'm moving towards standardizing on using the `Danbooru::Http` wrapper for `httprb` for a few reasons:

* We directly or indirectly pull in too many dependencies on too many different http libraries. We have `httprb`, `HTTParty`, `Mechanize`, `Faraday` (pulled in by the `oauth2` gem), `Net::Http::Persistent` (pulled in by mechanize`), `Net::Http` (the ruby standard http library, which mechanize and httparty are just wrappers around).
* Having this many HTTP libraries makes it hard to configure logging or proxies.
* HTTParty doesn't support persistent connections. Meaning that if we have to make multiple HTTP calls for a strategy, we have to establish a new connection for each call. This adds noticeable latency.
* Mechanize has a bug we have to work around (see `config/initializers/mechanize_patch.rb`) and anyway, it's just a wrapper for `Net::HTTP` and `Nokogiri` with automatic cookie handling, but usually we have to manually handle cookies anyway for caching purposes. Mechanize is also designed for HTML scraping, not for JSON APIs, so we end up needing to use another HTTP library anyway for non-HTML calls.
* `httprb` supports persistent connections and has a cleaner API than the other libraries. It doesn't have automatic support for cookies or Nokogiri, but this isn't too much trouble to do ourselves (notwithstanding the cookie redirect bug).